apsuite/commisslib/measure_tbt_data.py

```diff
@@ -0,0 +1,310 @@
+        trigpingh.nr_pulses = state.get('trigpingv_nrpulses',
+                                        prms.trigpingv_nrpulses)
+        dly = state.get('trigpingv_delay', prms.trigpingv_delay)
+        if dly is not None:
+            trigpingv.delay = dly
+
+    def get_magnets_strength(self):
+        """."""
+        return self.devices['pingh'].strength, self.devices['pingv'].strength
+
+    def set_magnets_strength(self, hkick=None, vkick=None):
+        """."""
+        pingh, pingv = self.devices['pingh'], self.devices['pingv']
+        if hkick is not None:
+            pingh.strength = hkick
+        else:
+            pingh.strength = self.params.hkick / 1e6
+        _time.sleep(0.5)
+        if vkick is not None:
+            pingv.strength = vkick
+        else:
+            pingv.strength = self.params.vkick / 1e6
+        _time.sleep(0.5)
+
+    def do_measurement(self):
+        """."""
+        currinfo = self.devices['currinfo']
+        init_timing_state = self.get_timing_state()
+        init_magnets_strength = self.get_magnets_strength()
+        current_before = currinfo.current()
+        self.prepare_timing()
+        self.set_magnets_strength()  # gets strengths from params
+        self.data['measurement_error'] = False  # error flag
+        try:
+            self.acquire_data()  # BPMs signals + relevant info are acquired
+                                 # such as timestamps tunes, stored current
+                                 # rf frequency, acq rate, nr samples, etc.
+        except Exception as e:
+            print(f'An error occurred during acquisition: {e}')
+            self.data['measurement_error'] = True
+        self.recover_timing_state(init_timing_state)
+        self.set_magnets_strength(init_magnets_strength)  # restore strengths
+        self.data['current_before'] = current_before
+        self.data['current_after'] = self.data.pop('stored_current')
+
+    def get_default_fname(self):
+        """."""
+        hkick, vkick = self.params.hkick, self.params.vkick
+        tm = self.data['timestamp']
+        fmt = '%Y-%m-%d-%H-%M-%S'
+        tmstp = _datetime.datetime.fromtimestamp(tm).strftime(fmt)
+        stg = f'hkick_{int(round(hkick)):3d}_vkick_{int(round(vkick)):3d}'
+        stg += f'_urad_{tmstp}'
+        return stg
+
+
+class TbTDataAnalysis(_AcqBPMsSignals):
+    """."""
+
+    def __init__(self, filename='', isonline=False):
+        """Analysis of linear optics using Turn-by-turn data."""
+        self.params = TbTDataParams()
+        self.isonline = isonline
+        self._ispost_mortem = False
+
+        self._fname = filename
+        self._trajx, self._trajy = None, None
+        self._trajsum = None
+
+        # load if fname, load method
+
+    @property
+    def fname(self):
+        """."""
+        return self._fname
+
+    @fname.setter
+    def fname(self, val):
+        self._fname = val
+
+    @property
+    def trajx(self):
```

Comment:
you should get all these properties from the data loaded from file. And remember to return a copy of them, to guarantee data integrity.